File: 1/2021.03.27/450.Delete Node in a BST/python/solution.py
class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        def get_min(root):
            while root.left:
                root = root.left
            return root

        if root:
            if root.val == key:
                if not root.left and not root.right:
                    return None
                if not root.left:
                    return root.right
                if not root.right:
                    return root.left
                min_node = get_min(root.right)
                root.val = min_node.val
                root.right = self.deleteNode(root.right, min_node.val)
            elif key < root.val:
                root.left = self.deleteNode(root.left, key)
            else:
                root.right = self.deleteNode(root.right, key)
            return root

File: 1/2021.03.27/450.Delete Node in a BST/java/solution.java
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root != null) {
            if (key == root.val) {
                if (root.left == null && root.right == null) {
                    return null;
                }
                if (root.left == null) {
                    return root.right;
                }
                if (root.right == null) {
                    return root.left;
                }
                TreeNode minNode = getMin(root.right);
                root.val = minNode.val;
                root.right = deleteNode(root.right, minNode.val);
            } else if (key < root.val) {
                root.left = deleteNode(root.left, key);
            } else {
                root.right = deleteNode(root.right, key);
            }
        }
        return root;
    }

    private TreeNode getMin(TreeNode root) {
        while (root.left != null) {
            root = root.left;
        }
        return root;
    }
}
